src/core/bpf-lsm.c

```diff
@@ -0,0 +1,375 @@
+                return supported = 0;
+        }
+
+        if (r == 0) {
+                log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                         "BPF LSM hook not enabled in the kernel");
+                return supported = 0;
+        }
+
+        /*
+         * Try loading bpf object with inner map to probe whether the BPF_PROG_TYPE_LSM program type and
+         * BTF are supported by kernel and if resource limits allow locking enough memory for bpf programs.
+         */
+        r = bpf_object_new(restrict_fs_hexdump_buffer, sizeof(restrict_fs_hexdump_buffer), &obj);
+        if (r < 0) {
+                log_error_errno(r, "Failed to create BPF object from hexdump buffer: %m");
+                return supported = 0;
+        }
+
+        /* Dummy map to satisfy the verifier */
+        inner_map_fd = bpf_create_map(BPF_MAP_TYPE_HASH, sizeof(uint32_t), sizeof(uint32_t), 128, 0);
+        if (inner_map_fd < 0) {
+                log_debug_errno(errno, "Failed to create BPF map");
+                return supported = 0;
+        }
+
+        r = bpf_object_set_inner_map_fd(obj, "cgroup_hash", inner_map_fd);
+        if (r < 0) {
+                log_debug_errno(r, "Failed to set inner map fd: %m");
+                return supported = 0;
+        }
+
+        r = bpf_object_load(obj);
+        if (r < 0) {
+                log_debug_errno(errno, "Failed to load LSM BPF program: %m");
+                return supported = 0;
+        }
+
+        return supported = 1;
+}
+
+int lsm_bpf_setup(void) {
+        _cleanup_(bpf_object_freep) struct bpf_object *obj = NULL;
+        struct bpf_program *prog = NULL;
+        struct bpf_link *link = NULL;
+        _cleanup_close_ int inner_map_fd = -1;
+        int outer_map_fd;
+        int r;
+
+        r = bpf_object_new(restrict_fs_hexdump_buffer, sizeof(restrict_fs_hexdump_buffer), &obj);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create BPF object from hexdump buffer: %m");
+
+        /* Dummy map to satisfy the verifier */
+        inner_map_fd = bpf_create_map(BPF_MAP_TYPE_HASH, sizeof(uint32_t), sizeof(uint32_t), 128, 0);
+        if (inner_map_fd < 0)
+                return log_error_errno(errno, "Failed to create BPF map: %m");
+
+        r = bpf_object_set_inner_map_fd(obj, "cgroup_hash", inner_map_fd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set inner map fd: %m");
+
+        r = bpf_object_load(obj);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to load LSM BPF program: %m");
+
+        /* Close dummy map */
+        inner_map_fd = safe_close(inner_map_fd);
+
+        r = bpf_object_find_program_by_title(obj, LSM_PROG_SECTION_NAME, &prog);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find LSM BPF program '%s': %m", LSM_PROG_SECTION_NAME);
+
+        link = bpf_program__attach_lsm(prog);
```

Comment:
This will attach the program to the LSM hook in all the cases even if the `RestrictFileSystems` feature is not used at all. Is that possible to implement a conditional logic that only (loads? and) attaches the program when that property is needed?

Reply:
`bpf_link` is leaked here: memory allocated in https://github.com/libbpf/libbpf/blob/master/src/libbpf.c#L9893 is not freed.
Also link fd is not closed and link stays attached after `lsm_bpf_setup` exits.
`bpf_link__destroy` should be used to close fd and free memory, e.g. in cleanup function https://github.com/systemd/systemd/pull/17655/commits/8326f7cc9640c5dd4a61335291fa9b292a893ae6

Also for bpf_link to stay attached, it must be stored. `Manager` seems to be a proper place for it.